Source/Core/Core/HW/EXI/BBA/TAPServerConnection.cpp

```diff
@@ -0,0 +1,356 @@
+  return true;
+}
+
+void TAPServerConnection::RecvStart()
+{
+  m_read_enabled.Set();
+}
+
+void TAPServerConnection::RecvStop()
+{
+  m_read_enabled.Clear();
+}
+
+bool TAPServerConnection::SendAndRemoveAllHDLCFrames(std::string* send_buf)
+{
+  while (!send_buf->empty())
+  {
+    const std::size_t start_offset = send_buf->find(0x7E);
+    if (start_offset == std::string::npos)
+    {
+      break;
+    }
+    const std::size_t end_sentinel_offset = send_buf->find(0x7E, start_offset + 1);
+    if (end_sentinel_offset == std::string::npos)
+    {
+      break;
+    }
+    const std::size_t end_offset = end_sentinel_offset + 1;
+    const std::size_t size = end_offset - start_offset;
+
+    const u8 size_bytes[2] = {static_cast<u8>(size), static_cast<u8>(size >> 8)};
+    if (send(m_fd, reinterpret_cast<const char*>(size_bytes), 2, SEND_FLAGS) != 2)
+    {
+      ERROR_LOG_FMT(SP1, "SendAndRemoveAllHDLCFrames(): could not write size field");
+      return false;
+    }
+    const int written_bytes =
+        send(m_fd, send_buf->data() + start_offset, static_cast<int>(size), SEND_FLAGS);
+    if (u32(written_bytes) != size)
+    {
+      ERROR_LOG_FMT(SP1,
+                    "SendAndRemoveAllHDLCFrames(): expected to write {} bytes, instead wrote {}",
+                    size, written_bytes);
+      return false;
+    }
+    *send_buf = send_buf->substr(end_offset);
+  }
+  return true;
+}
+
+bool TAPServerConnection::SendFrame(const u8* frame, u32 size)
+{
+  INFO_LOG_FMT(SP1, "SendFrame {}\n{}", size, ArrayToString(frame, size, 0x10));
+
+  // On Windows, the data pointer is of type const char*; on other systems it is
+  // of type const void*. This is the reason for the reinterpret_cast here and
+  // in the other send/recv calls in this file.
+  const u8 size_bytes[2] = {static_cast<u8>(size), static_cast<u8>(size >> 8)};
+  if (send(m_fd, reinterpret_cast<const char*>(size_bytes), 2, SEND_FLAGS) != 2)
+  {
+    ERROR_LOG_FMT(SP1, "SendFrame(): could not write size field");
+    return false;
+  }
+  const int written_bytes =
+      send(m_fd, reinterpret_cast<const char*>(frame), static_cast<ws_ssize_t>(size), SEND_FLAGS);
+  if (u32(written_bytes) != size)
+  {
+    ERROR_LOG_FMT(SP1, "SendFrame(): expected to write {} bytes, instead wrote {}", size,
+                  written_bytes);
+    return false;
+  }
+  return true;
+}
+
+void TAPServerConnection::ReadThreadHandler()
+{
+  enum class ReadState
+  {
+    SIZE,
+    SIZE_HIGH,
+    DATA,
+    SKIP,
+  };
+  ReadState read_state = ReadState::SIZE;
+
+  std::size_t frame_bytes_received = 0;
+  std::size_t frame_bytes_expected = 0;
+  std::string frame_data;
+
+  while (!m_read_shutdown.IsSet())
+  {
+    fd_set rfds;
+    FD_ZERO(&rfds);
+    FD_SET(m_fd, &rfds);
+
+    timeval timeout;
+    timeout.tv_sec = 0;
+    timeout.tv_usec = 50000;
+    if (select(m_fd + 1, &rfds, nullptr, nullptr, &timeout) <= 0)
+      continue;
+
```

Comment:
This might silent errors, you can continue on `0`/`timeout` but if an error occurs (< 0) it should be handled and logged properly.